src/storage/heap_file.c

```diff
@@ -21248,55 +21239,59 @@ heap_delete_home (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context, boo
 	  use_optimization = false;
 	}
 
+#if !defined(NDEBUG)
+      if (is_adjusted_size_big)
+	{
+	  /* not exactly necessary, but we'll be able to compare sizes */
+	  adjusted_size = context->home_recdes.length - or_mvcc_header_size_from_flags (mvcc_flags)
+	    + OR_MVCC_MAX_HEADER_SIZE;
+	}
+#endif
+
       /* Build the new record. */
       HEAP_SET_RECORD (&built_recdes, IO_DEFAULT_PAGE_SIZE + OR_MVCC_MAX_HEADER_SIZE, 0, REC_UNKNOWN,
 		       PTR_ALIGN (data_buffer, MAX_ALIGNMENT));
       if (use_optimization)
 	{
-	  /* Most common case, computes offset of DELID. */
-	  if (mvcc_flags & OR_MVCC_FLAG_VALID_PREV_VERSION)
-	    {
-	      /* Have previous version. */
-	      offset = header_size - sizeof_log_lsa;
-	    }
-	  else
-	    {
-	      /* Don't have previous version. */
-	      offset = header_size;
-	    }
+	  char *start_p;
+
+	  delid_offset = OR_MVCC_DELETE_ID_OFFSET (mvcc_flags);
+
+	  build_recdes_data = start_p = built_recdes.data;
 
-	  build_recdes_data = built_recdes.data;
 	  /* Copy up to MVCC DELID first. */
-	  memcpy (build_recdes_data, context->home_recdes.data, offset);
-	  build_recdes_data += offset;
+	  memcpy (build_recdes_data, context->home_recdes.data, delid_offset);
+	  build_recdes_data += delid_offset;
 
 	  /* Sets MVCC DELID flag, overwrite first four bytes. */
 	  repid_and_flag_bits |= (OR_MVCC_FLAG_VALID_DELID << OR_MVCC_FLAG_SHIFT_BITS);
-	  OR_PUT_INT (built_recdes.data, repid_and_flag_bits);
+	  OR_PUT_INT (start_p, repid_and_flag_bits);
+
+	  /* Sets the MVCC DELID. */
+	  OR_PUT_BIGINT (build_recdes_data, &mvcc_id);
+	  build_recdes_data += OR_MVCC_DELETE_ID_SIZE;
 
+	  /* Copy remaining data. */
+#if !defined(NDEBUG)
 	  if (mvcc_flags & OR_MVCC_FLAG_VALID_PREV_VERSION)
 	    {
-	      /* Sets the previous version. */
-	      memcpy (build_recdes_data + OR_MVCCID_SIZE, context->home_recdes.data + offset, sizeof_log_lsa);
-
-	      /* Sets the MVCC DELID. */
-	      OR_PUT_BIGINT (build_recdes_data, &mvcc_id);
-	      build_recdes_data += (OR_MVCCID_SIZE + sizeof_log_lsa);
+	      /* Check that we need to copy from offset of LOG LSA up to the end of the buffer. */
+	      assert (delid_offset == OR_MVCC_PREV_VERSION_LSA_OFFSET (mvcc_flags));
 	    }
 	  else
 	    {
-	      /* Sets the MVCC DELID. */
-	      OR_PUT_BIGINT (build_recdes_data, &mvcc_id);
-	      build_recdes_data += OR_MVCCID_SIZE;
+	      /* Check that we need to copy from end of MVCC header up to the end of the buffer. */
+	      assert (delid_offset == or_mvcc_header_size_from_flags (mvcc_flags));
 	    }
+#endif
 
-	  /* Copy behind MVCC DELID. */
-	  memcpy (build_recdes_data, context->home_recdes.data + header_size,
-		  context->home_recdes.length - header_size);
+	  memcpy (build_recdes_data, context->home_recdes.data + delid_offset,
+		  context->home_recdes.length - delid_offset);
 	  built_recdes.length = adjusted_size;
```

Comment:
Should be... when optimization is used, adjusted_size = recdes.length + OR_MVCCID_SIZE, if OR_MVCC_FLAG_VALID_DELID is not set (the new record length = old record length + 8 bytes), otherwise adjusted_size = recdes.length  (the new record length = old record length).